src/Chainweb/Version/Utils.hs

```diff
@@ -0,0 +1,434 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- |
+-- Module: Chainweb.Version.Utils
+-- Copyright: Copyright © 2020 Kadena LLC.
+-- License: MIT
+-- Maintainer: Lars Kuhtz <lars@kadena.io>
+-- Stability: experimental
+--
+-- Utilities for computing BlockHeights, CutHeights and BlockCounts based on
+-- chainweb versions and chainweb graphs.
+--
+module Chainweb.Version.Utils
+(
+-- * Chain Ids
+  randomChainId
+, someChainId
+
+-- * Chain Graph Properties By Block Height
+, chainGraphs
+, chainGraphsAt
+, lastGraphChange
+, nextGraphChange
+, chainCountAt
+, diameterAt
+, degreeAt
+, chainIdsAt
+, randomChainIdAt
+, someChainIdAt
+, blockCountAt
+, globalBlockCountAt
+, globalBlockRateAt
+, isGraphChange
+
+-- * Chain Graph Properties By Cut Height
+, avgCutHeightAt
+, chainGraphsByCutHeight
+, chainGraphAtCutHeight
+, chainCountAtCutHeight
+, diameterAtCutHeight
+, degreeAtCutHeight
+, blockCountAtCutHeight
+, avgBlockHeightAtCutHeight
+, lastGraphChangeByCutHeight
+, nextGraphChangeByCutHeight
+
+-- * Expected Block and Cut Heights
+, expectedBlockHeightAfterSeconds
+, expectedCutHeightAfterSeconds
+, expectedBlockCountAfterSeconds
+, expectedGlobalBlockCountAfterSeconds
+) where
+
+import Chainweb.BlockHeight
+import Chainweb.Time
+
+import Data.Foldable
+import qualified Data.HashSet as HS
+import qualified Data.Map.Strict as M
+
+import GHC.Stack
+
+import Numeric.Natural
+
+import System.Random
+
+-- internal modules
+
+import Chainweb.Graph
+import Chainweb.Utils
+import Chainweb.Version
+
+-- -------------------------------------------------------------------------- --
+--  Utils
+
+limitHeight :: BlockHeight -> M.Map BlockHeight a -> M.Map BlockHeight a
+limitHeight h = M.takeWhileAntitone (h >=)
+{-# INLINE limitHeight #-}
+
+atHeight :: HasCallStack => BlockHeight -> M.Map BlockHeight a -> a
+atHeight h = snd . fromJuste . M.lookupLE h
+{-# INLINE atHeight #-}
+
+atCutHeight :: HasCallStack => CutHeight -> M.Map CutHeight a -> a
+atCutHeight h = snd . fromJuste . M.lookupLE h
+{-# INLINE atCutHeight #-}
+
+-- -------------------------------------------------------------------------- --
+-- Chain Graph Properties By Block Height
+
+-- | BlockHeight intervals for the chain graphs of a chainweb version
+--
+-- Post-condition:
+--
+-- @
+-- 0 == minimum $ M.keys $ chainGraphs v
+-- @
+--
+chainGraphs
```

Comment:
See note in `Version`, seems like we should standardize on `Map` as otherwise we're creating a `NonEmpty`, destructing it and creating a `Map` for every operation in here.

Reply:
Yeah, it's possibly inefficient. However, I think the implementation is optimized and fuses well. This module uses `Map` not as a data structure but more as an efficient search algorithm.

After writing a lot of specialized code for dealing with intervals I found that using map was just the fastest and safest way to get to a working implementation quickly.

Map handles a lot of nasty corner cases of the required traversals. The explicit implementations in my first approach sometimes got pretty complex due to those corner cases and there is some chance that Map isn't that bad in comparison, even performance wise.

But there is some chance that after profiling we'll have to revisit this module and replace the use of Map by more efficient implementations.

Reply:
After thinking a bit more about it, I think, a good solution could be, to keep using `Data.List.NonEmpty` in `Chainweb.Version` for specifying the graphs and to memoize the result of this function for the three production chainweb versions `mainet01`, `development`, and `testnet04` here.

Something like:

```Haskell
chainGraphs :: HasChainwebVersion v => v -> M.Map BlockHeight ChainGraph
chainGraphs v = case _chainwebVersion v of
    Mainnet01 -> mainnet01GraphMap
    Testnet04 -> testnet04GraphMap
    Development -> developmentGraphMap
    x -> M.fromList . toList $ chainwebGraphs x

mainnet01GraphMap :: M.Map BlockHeight ChainGraph
mainnet01GraphMap = M.fromList . toList $ chainwebGraphs Mainnet01
{-# NOINLINE mainnet01GraphMap #-}

testnet04GraphMap :: M.Map BlockHeight ChainGraph
testnet04GraphMap = M.fromList . toList $ chainwebGraphs Testnet04
{-# NOINLINE testnet04GraphMap #-}

developmentGraphMap :: M.Map BlockHeight ChainGraph
developmentGraphMap = M.fromList . toList $ chainwebGraphs Development
{-# NOINLINE developmentGraphMap #-}
```

That has the benefit of keeping the specification in `Chainweb.Version` simple and maintainable and providing an efficient Map in this module. It would also allow us to do further optimization (e.g. using arrays/vector instead of Map) without "disturbing" `Chainweb.Version`.